File: CPP/Grid/Magic Squares In Grid.cpp
class Solution {
public:
    bool isMagic(vector<vector<int>>&grid, int rows, int cols){
        set<int>st;
        for(int i=rows;i<rows+3;i++){
            for(int j=cols;j<cols+3;j++){
                int num=grid[i][j];
                if(num<1 || num>9 ||st.find(num)!=st.end()){
                    return false;
                }
                st.insert(num);
            }
        }
        //diagonal sum
        int diag1=grid[rows][cols]+grid[rows+1][cols+1]+grid[rows+2][cols+2];
        int diag2=grid[rows][cols+2]+grid[rows+1][cols+1]+grid[rows+2][cols];
        if(diag1!=diag2)return false;
        //row sum
        for(int i=rows;i<rows+3;i++){
            if(diag1!=grid[i][cols]+grid[i][cols+1]+grid[i][cols+2]){
                return false;
            }
        }
        //col sum
        for(int i=cols;i<cols+3;i++){
            if(diag1!=grid[rows][i]+grid[rows+1][i]+grid[rows+2][i]){
                return false;
            }
        }
        return true;
    }
    int numMagicSquaresInside(vector<vector<int>>& grid) {
        int rz=grid.size();
        int cz=grid[0].size();
        if(rz<3 || cz<3)return false;
        int ans=0;
        for(int rows=0;rows+2<rz;rows++){
            for(int cols=0;cols+2<cz;cols++){
                if(isMagic(grid,rows,cols)){
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: Java/2D Array(Matrix)/Magic Squares In Grid.java
class Solution {
    private boolean check(int [][]grid, int r, int c){
        boolean[] seen = new boolean[10]; // Use size 10 : bcz seen has indices 0..8 
                                        // You access seen[num] where num can be 9 → ArrayIndexOutOfBoundsException
        for(int i=r;i<r+3; i++){
            for(int j=c;j<c+3;j++){
                int num = grid[i][j];
                if (num <= 0 || num > 9 || seen[num]==true) {
                    return false;
                }
                seen[num]=true;
            }
        }

        // fist diagonal sum
        int diag1 = grid[r][c] + grid[r + 1][c + 1] + grid[r + 2][c + 2];

        // second diagonal sum
        int diag2 = grid[r][c + 2] + grid[r + 1][c + 1] + grid[r + 2][c];

        if (diag1 != diag2)
            return false;

        // Check row sum
        for (int i = r; i < r + 3; i++) {
            if (diag1 != grid[i][c] + grid[i][c + 1] + grid[i][c + 2])
                return false;
        }

        // Check col sum
        for (int i = c; i < c + 3; i++) {
            if (diag1 != grid[r][i] + grid[r + 1][i] + grid[r + 2][i])
                return false;
        }
        return true;
    }

    public int numMagicSquaresInside(int[][] grid) {
        int row = grid.length;
        int col = grid[0].length;

        if (row < 3 || col < 3)
            return 0;

        int ans = 0;
        for (int i = 0; i <= row - 3; i++) {
            for (int j = 0; j <= col - 3; j++) {
                if (check(grid, i, j))
                    ans++;
            }
        }
        return ans;
    }
}
